lib/util/tmp.js

```diff
@@ -22,16 +23,35 @@ function mktmpdir (cache, opts) {
 }
 
 module.exports.withTmp = withTmp
+
 function withTmp (cache, opts, cb) {
   if (!cb) {
     cb = opts
     opts = null
   }
   opts = TmpOpts(opts)
-  return BB.using(mktmpdir(cache, opts).disposer(rimraf), cb)
+
+  const disposer = (target, result, shouldThrow = false) => {
+    const returnResult = () => {
+      if (shouldThrow) {
+        throw result
+      }
+      return result
+    }
+    return rimraf(target)
+      .then(() => returnResult())
+      .catch(() => returnResult())
+  }
+
+  return mktmpdir(cache, opts).then((target) =>
+    cb(target)
```

Comment:
`cb` is **optionally** a Promise but per docs it can also be a simple function, ref: https://github.com/npm/cacache#--tmpwithtmpcache-opts-cb---promise

 so I guess we either need to document that breaking change of behavior or add some extra logic to handle that here

Reply:
Sorry, I might have missed it, but I don't think `cb` is optional here. 

This basically calls `mktmpdir` then passes the new directory to `cb` which is a function that is passed in. And once `cb` resolves, the disposer is ran.

Reply:
I think we should just update the README to remove this part:

> -- that is, if you return a Promise from cb()

Reply:
Removed this part of the doc.